File: app/src/main/java/dreammaker/android/expensetracker/view/adapter/AccountsListAdapter.java
package dreammaker.android.expensetracker.view.adapter;

import android.content.Context;
import android.text.SpannableString;
import android.text.SpannableStringBuilder;
import android.text.Spanned;
import android.text.TextUtils;
import android.text.style.ForegroundColorSpan;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.DiffUtil;
import dreammaker.android.expensetracker.R;
import dreammaker.android.expensetracker.database.model.AccountDisplayModel;

public class AccountsListAdapter extends BaseClickableItemRecyclerViewListAdapter<AccountDisplayModel, AccountsListAdapter.AccountViewHolder>{

    private static DiffUtil.ItemCallback<AccountDisplayModel> DIFF_CALLBACK = new DiffUtil.ItemCallback<AccountDisplayModel>() {
        @Override
        public boolean areItemsTheSame(@NonNull AccountDisplayModel oldItem, @NonNull AccountDisplayModel newItem) {
            return oldItem.getId() == newItem.getId();
        }

        @Override
        public boolean areContentsTheSame(@NonNull AccountDisplayModel oldItem, @NonNull AccountDisplayModel newItem) {
            return oldItem.equals(newItem);
        }
    };

    public AccountsListAdapter(@NonNull Context context) {
        super(context, DIFF_CALLBACK,true);
    }

    @Override
    protected AccountViewHolder onCreateViewHolder(@NonNull LayoutInflater inflater, @NonNull ViewGroup parent, int viewType) {
        AccountViewHolder vh = new AccountViewHolder(inflater.inflate(R.layout.list_item_two_lines,parent,false));
        return vh;
    }

    @Override
    public void onBindViewHolder(@NonNull AccountViewHolder holder, int position) {
        holder.bind(getItem(position));
    }

    @Override
    protected List<AccountDisplayModel> onFilter(@Nullable CharSequence constraint) {
        final List<AccountDisplayModel> original = getOriginalItems();
        final List<AccountDisplayModel> current = getCurrentList();
        if (null == original || original.isEmpty() || TextUtils.isEmpty(constraint)) return original;
        if (null == current || current.isEmpty()) return null;
        String phrase = constraint.toString();
        ArrayList<AccountDisplayModel> values = new ArrayList<>();
        for (AccountDisplayModel a : current) {
            String name = a.getName();
            if (name.contains(phrase)) {
                values.add(a);
            }
        }
        return values;
    }

    public static class AccountViewHolder extends BaseClickableItemViewHolder<AccountDisplayModel> {

        TextView line1;
        TextView line2;
        View btnOptions;

        public AccountViewHolder(@NonNull View itemView) {
            super(itemView);
            line1 = findViewById(R.id.line1);
            line2 = findViewById(R.id.line2);
            btnOptions = findViewById(R.id.options);
            itemView.setOnClickListener(this);
            btnOptions.setOnClickListener(this);
        }

        @Override
        public void bind(@Nullable AccountDisplayModel item, @Nullable Object payload) {
            if (null != item) {
                line1.setText(item.getName());
                BigDecimal balance = item.getBalance();
                SpannableString txtBalance = SpannableString.valueOf(balance.toPlainString());
                int txtBalanceColor = balance.compareTo(BigDecimal.ZERO) > 0 ? getColor(R.color.text_green) : getColor(R.color.text_red);
                txtBalance.setSpan(new ForegroundColorSpan(txtBalanceColor),0,txtBalance.length(), Spanned.SPAN_INCLUSIVE_EXCLUSIVE);
                SpannableStringBuilder builder = new SpannableStringBuilder();
                builder.append(getString(R.string.label_total_balance)).append(": ").append(txtBalance);
                line2.setText(builder);
            }
            else {
                // TODO: show loading

            }
        }
    }
}

File: app/src/main/java/dreammaker/android/expensetracker/ui/account/accountlist/AccountsListAdapter.kt
package dreammaker.android.expensetracker.ui.account.accountlist

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import dreammaker.android.expensetracker.databinding.AccountListItemBinding
import dreammaker.android.expensetracker.util.BaseSelectableItemListAdapter
import dreammaker.android.expensetracker.util.SelectableViewHolder
import dreammaker.android.expensetracker.util.getBalanceText
import rahulstech.android.expensetracker.domain.model.Account

class AccountViewHolder(
    val binding: AccountListItemBinding
): SelectableViewHolder<Long>(binding.root) {

    // NOTE: without RecyclerView.Adapter.setHasStableId(true) RecyclerView.ViewHolder.itemId is always invalidd
    private var stableItemId: Long? = null

    override fun getSelectionKey(): Long? = stableItemId

    fun bind(data: Account?, selected: Boolean) {
        stableItemId = data?.id
        if (null == data) {
            binding.name.text = null
            binding.balance.text = null
            binding.root.isActivated = false
        }
        else {
            binding.name.text = data.name
            binding.balance.text = data.getBalanceText(context)
            binding.root.isActivated = selected
        }
    }
}

private val callback = object: DiffUtil.ItemCallback<Account>() {
    override fun areItemsTheSame(oldItem: Account, newItem: Account): Boolean =
        oldItem.id == newItem.id

    override fun areContentsTheSame(oldItem: Account, newItem: Account): Boolean =
        oldItem == newItem
}

class AccountsAdapter: BaseSelectableItemListAdapter<Account, Long, AccountViewHolder>(callback) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AccountViewHolder {
        val inflate = LayoutInflater.from(parent.context)
        val binding = AccountListItemBinding.inflate(inflate, parent, false)
        return AccountViewHolder(binding).apply {
            attachItemClickListener {  vh, v -> handleItemClick(vh as AccountViewHolder,v) }
            attachItemLongClickListener { vh, v -> handleItemLongClick(vh as AccountViewHolder,v) }
        }
    }

    override fun onBindViewHolder(holder: AccountViewHolder, position: Int) {
        val item = getItem(position)
        holder.bind(item,isSelected(item.id))
    }

    override fun getItemId(position: Int): Long = getItem(position)?.id ?: RecyclerView.NO_ID

    override fun getSelectionKeyAtPosition(position: Int): Long? {
        val id = getItemId(position)
        if (id == RecyclerView.NO_ID) return null
        return id
    }
}